crates/wasi-common/src/snapshots/preview_1.rs

```diff
@@ -1029,11 +1111,21 @@ impl wasi_snapshot_preview1::WasiSnapshotPreview1 for WasiCtx {
         buf: &GuestPtr<'a, u8>,
         buf_len: types::Size,
     ) -> Result<(), Error> {
-        let mut buf = buf
-            .as_array(buf_len)
-            .as_slice_mut()?
-            .expect("cannot use with shared memories; see https://github.com/bytecodealliance/wasmtime/issues/5235 (TODO)");
-        self.random.try_fill_bytes(buf.deref_mut())?;
+        let buf = buf.as_array(buf_len).as_unsafe_slice_mut()?;
+        if buf.is_shared_memory() {
+            // If the Wasm memory is shared, copy to an intermediate buffer to
+            // avoid Rust unsafety (i.e., the called function could rely on
+            // `&mut [u8]`'s exclusive ownership which is not guaranteed due to
+            // potential access from other threads).
+            let mut tmp = vec![0; buf.len().min(MAX_SHARED_BUFFER_SIZE)];
+            self.random.try_fill_bytes(&mut tmp)?;
+            buf.copy_from_slice(&tmp)?;
```

Comment:
Since `random_get` fills the entire buffer I think this will need to `loop` or otherwise this will return an error since `tmp` may be shorter than `buf`

Reply:
You might want to take a look at 652988f since I changed the `GuestPtr` API slightly.